kubelet: add a generic pod lifecycle event generator

yujuhong opened the pull request:
This change introduces pod lifecycle event generator (PLEG), and adds a generic
PLEG. The generic PLEG relies on relisting to discover container events, and is
container-runtime-agnostic. Both docker and rkt are changed to use generic
PLEG.

yujuhong committed:
kubelet: add a generic pod lifecycle event generator

This change introduces pod lifecycle event generator (PLEG), and adds a generic
PLEG. The generic PLEG relies on relisting to discover container events, and is
container-runtime-agnostic. Both docker and rkt are changed to use generic
PLEG.

yujuhong commented:
This is part of #12802 to improve kubelet's performance. I decided to split #12829 in to a few parts for easier reviewing. The plan is to
  1. Add a generic PLEG for all container runtimes. Switch docker and rkt to use it.
  2. Add a docker PLEG which uses docker event stream. Switch docker to use it.

This PR does (1), and I am working on (2).

Other parallel tasks to reduce sync period includes:
* Instruct pod workers to requeue themselves on sync error. I have a local patch for this.
* Move probing out of sync loop (#12866).
* Handle pastActiveDeadline check. I will address this later.


What to expect from this PR:
  - kubelet should react to container termination events more quickly as the generic PLEG relists every ~3 seconds.
  - Temporarily rise in number of syncs because generic PLEG sends sync events even on container starts as well. At the same time, we still have the same 10s period sync period. 
      - I don't expect this to affect the performance of kubelet too much, since the recent PR #13003 has lowered the sync significantly.
      - This would get better when we add the docker PLEG and/or increase the sync period.

yujuhong commented:
cc @kubernetes/goog-node

timstclair commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -0,0 +1,149 @@
+// Returns a channel from which the subscriber can recieve PodLifecycleEvent
+// events.
+func (g *GenericPLEG) Watch() chan *PodLifecycleEvent {
+	return g.eventChannel
```
With a single channel, it means only a single component can be listening to the PLEs, right? It would be useful for the ProbeManager (work in progress by me) to be able to listen for the Container{Started,Stopped} events. What if each call to Watch created a new channel, which the "main" channel cloned events into? It would add a bit more complexity around the bookkeeping of closing channels & preventing channels from backing up though.

yujuhong replied:
Yes, I had a TODO to extend this to multiple channels, but it's still a TODO :)
I think multi-channel support can wait a little bit, and help is always welcome.

yujuhong commented:
Rebased and a friendly ping.

yujuhong commented:
@k8s-bot ok to test.

yujuhong commented:
@k8s-bot ok to test.

yifan-gu commented on pkg/kubelet/container/fake_runtime.go:
```diff
@@ -149,7 +151,11 @@ func (f *FakeRuntime) GetPods(all bool) ([]*Pod, error) {
-	return f.PodList, f.Err
+	if all {
+		return f.AllPodList, f.Err
+	} else {
```
unnecessary else

yujuhong replied:
Removed.

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -1919,8 +1942,27 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan PodUpdate, handler SyncHandl
 			glog.Errorf("Kubelet does not support snapshot update")
 		}
+	case e := <-plegCh:
+		if pod, ok := kl.podManager.GetPodByUID(e.ID); ok {
```
nit: I think swapping the order can let us have less indent here, e.g.
```go
pod, ok := kl.podManager.GetPodByUID(e.ID)
if !ok {
    break
}
glog.V(2).Infof(SyncLoop (PLEG): %q, event: %#v", kubeletUtil.FormatPodName(pod), e)
...
```

yujuhong replied:
Done. I am all for less indent :)

yujuhong commented:
Rebased and addressed all comments.
@dchen1107, @yifan-gu, @vishh

yifan-gu commented on pkg/kubelet/kubelet.go:
```diff
@@ -1935,8 +1959,29 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan PodUpdate, handler SyncHandl
+		handler.HandlePodSyncs([]*api.Pod{pod})
 	case <-time.After(kl.resyncInterval):
 		// Periodically syncs all the pods and performs cleanup tasks.
+		//
```
Not in this PR, but like I said in #13990 We'd better use a ticker here, otherwise if we keep getting events from plegCh, then we won't be able to resync. (unless this is the preferred behavior)

yujuhong replied:
Done.

yifan-gu commented:
LGTM except one comment :)

yujuhong commented:
> LGTM except one comment :)

Addressed the comment. Will also update it in the docker variant. Thanks for reviewing :)

yujuhong commented:
@vishh, ping!

vishh commented on pkg/kubelet/pleg/pleg.go:
```diff
@@ -0,0 +1,52 @@
+	PodSync PodLifeCycleEventType = "PodSync"
+)
+
+// PodLifecycleEvent is an event reflects the change of the pod state.
```
nit: an event *that* reflects the.

yujuhong replied:
Done

vishh commented on pkg/kubelet/pod_manager.go:
```diff
@@ -171,6 +172,15 @@ func (pm *basicPodManager) getAllPods() []*api.Pod {
+func (pm *basicPodManager) GetPodByUID(uid types.UID) (*api.Pod, bool) {
+	pm.lock.RLock()
+	defer pm.lock.RUnlock()
+	pod, ok := pm.podByUID[uid]
```
Why not return `nil` when not found?

yujuhong replied:
We can. I think we used to return an `api.Pod` object (not a pointer), so it wasn't an option. Now that every function in this file has the same return format, we may want to change them in a separate pr.

vishh replied:
Acknowledged. Wanna file an issue for that? :)

vishh commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -0,0 +1,153 @@
+limitations under the License.
+*/
+
+package pleg
```
What is a pleg? I know it. May be add some package comments.

yujuhong replied:
Done. Added "doc.go"

vishh commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -0,0 +1,153 @@
+
+// GenericPLEG is an extremely simple generic PLEG that relies solely on
+// periodic listing to discover container changes. It should be be used
+// as temporary replacement or container runtimes do not support a proper
```
s/or/for

yujuhong replied:
Done

vishh commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -0,0 +1,153 @@
+var _ PodLifecycleEventGenerator = &GenericPLEG{}
+
+func NewGenericPLEG(runtime kubecontainer.Runtime, channelCapacity int,
+	relistPeriod time.Duration) *GenericPLEG {
```
Why not return `PodLifecycleEventGenerator`?

yujuhong replied:
This would make unit test unnecessarily complicated with type assertions.

vishh replied:
I assume you are referring to unit tests of this object? If thats true, why can't unit tests not use this factory method?

yujuhong replied:
What's the advantage of returning `PodlifecycleEventGenerator`?

vishh commented:
Overall LGTM @yujuhong. Just a few nits.

yujuhong commented:
Rebased to include the ContainerID changes, and addressed the comments.

yujuhong commented:
Rebased.

vishh commented:
@yujuhong: Just one optional nit and a mandatory rebase. 
LGTM.

wojtek-t commented:
@yujuhong - please rebase

yujuhong commented:
Rebased.

yujuhong committed:
Add Status in the runtime Container type

This is necessary for the generic PLEG to distinguish container events.

yujuhong commented:
@kubernetes/goog-node, I kept the original commit, but rewrote the entire PR in the following commits. The reason being that I still want the container event type which we can use in the future (for pod cache, etc). Deriving that information from calling GetPods is not possible. I decided to augment the container runtime interface to include this information. I believe it makes this PR much cleaner too. Please review the entire PR (not separate commits) because of the rewrite. By the way, I still need to fix the tests, so just ignore them for now.

@yifan-gu, a6d8bff9238b199113c881a6d321cd417a82cca6 changes the `Container` type. I modified rkt to return the information. Please voice your concerns!

yujuhong commented:
The PR is ready for review. @vishh @yifan-gu

yujuhong committed:
Adjust the sync/backoff period

Set resyncInterval to one minute now that we rely on the generic pleg to trigger
pod syncs on container events. When there is an error during syncing, pod
workers need to wake up sooner to retry. Set the sync error backoff period to
10 second in this case.

yifan-gu commented:
LGTM

pmorie commented:
@yujuhong I know there are folks that want this.  It looks like this PR got LGTM and then needed rebase.  Do you have availability to rebase and finish this out?

cc @sdminonne

yujuhong commented:
Rebased.

@vishh, are you ok with the new version?

vishh commented:
Reviewing now.

On Wed, Nov 11, 2015 at 1:00 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> Rebased.
>
> @vishh <https://github.com/vishh>, are you ok with the new version?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13571#issuecomment-155907966>
> .
>

yujuhong commented:
@k8s-bot test this please

vishh commented on pkg/kubelet/pleg/doc.go:
```diff
@@ -0,0 +1,19 @@
+limitations under the License.
+*/
+
+// Package pleg contains types and a generic implementation of the pod
```
nit: Can we add more information here on how the generic pleg works?

vishh replied:
nvm. I see the comments in `generic.go`.

vishh commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -0,0 +1,141 @@
+// garbage collector is implemented to work with such situtations. However, to
+// guarantee that kubelet can handle missing container events, it is
+// recommended to set the relist period short and have an auxiliary, longer
+// periodic sync in kubelet as the safety net.
```
What if we update other internal modules to use a PLEG based cache instead of reaching out the runtime directly? Is this something that you are already planning on doing?

yujuhong replied:
Yes, but instead of providing a cache with only output of ListContainers, I plan to add a cache that would store the inspect results as well. See #12619 for details.

yujuhong replied:
Yes, as we discussed in the team meeting, that's the next step :-)

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -2112,6 +2132,25 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan kubetypes.PodUpdate, handler
+			glog.V(4).Infof("SyncLoop (PLEG): ignore irrelevant event: %#v", e)
+			break
+		}
+		glog.V(2).Infof("SyncLoop (PLEG): %q, event: %#v", kubeletutil.FormatPodName(pod), e)
```
Our default logging level is set to `v2`. Should we lower this?

yujuhong replied:
Why do we want to lower this? I think it should be consistent with the logging level of other SyncLoop events (e.g., line 2123).

vishh replied:
Do we need this information most of the times? I'd expect to see only useful logs with V(2). What do you think will be the frequency of this log?

yujuhong replied:
This would only happen if there is a container event, which would be useful to know IMO. I think it would be okay as V(3) though. How about I do a followup PR to change logging levels of all messages in this function?

vishh replied:
SGTM.

On Thu, Nov 12, 2015 at 2:08 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/13571#discussion_r44722224>
> :
>
> > @@ -2112,6 +2132,25 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan kubetypes.PodUpdate, handler
> >  			// TODO: Do we want to support this?
> >  			glog.Errorf("Kubelet does not support snapshot update")
> >  		}
> > +	case e := <-plegCh:
> > +		// Filter out started events since we don't use them now.
> > +		if e.Type == pleg.ContainerStarted {
> > +			break
> > +		}
> > +		pod, ok := kl.podManager.GetPodByUID(e.ID)
> > +		if !ok {
> > +			// If the pod no longer exists, ignore the event.
> > +			glog.V(4).Infof("SyncLoop (PLEG): ignore irrelevant event: %#v", e)
> > +			break
> > +		}
> > +		glog.V(2).Infof("SyncLoop (PLEG): %q, event: %#v", kubeletutil.FormatPodName(pod), e)
>
> This would only happen if there is a container event, which would be
> useful to know IMO. I think it would be okay as V(3) though. How about I do
> a followup PR to change logging levels of all messages in this function?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13571/files#r44722224>.
>

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -2112,6 +2132,25 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan kubetypes.PodUpdate, handler
+		}
+		glog.V(2).Infof("SyncLoop (PLEG): %q, event: %#v", kubeletutil.FormatPodName(pod), e)
+		// Force the container runtime cache to update.
+		if err := kl.runtimeCache.ForceUpdateIfOlder(time.Now()); err != nil {
```
Why is this needed? Doesn't PLEG already handle relisting?

yujuhong replied:
PLGE doesn't populate this runtime cache shared by the pod workers. As mentioned in another comment, I plan to add a more general purpose cache that would eliminate the need of `docker inspect` as well. I can, of course, let PLEG populate this runtime cache first, but I feel like the benefits are marginal.

vishh replied:
Acknowledged!

vishh commented on pkg/kubelet/dockertools/convert.go:
```diff
@@ -27,6 +28,19 @@ import (
 // (kubecontainer) types.
 
+func mapStatus(status string) kubecontainer.ContainerStatus {
+	// Parse the status string in docker.APIContainers. This could break when
```
Why do you think this would break?

yujuhong replied:
This *might* break because we are parsing a human readable string. On the other hand, it should be easy to verify when upgrading docker.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -110,6 +111,18 @@ const (
 	etcHostsPath = "/etc/hosts"
+
+	// Capacity of the channel for recieving pod lifecycle events. This number
+	// is a bit arbitrary and may be adjusted in the future.
```
Can this be proportional to the machine capacity?

yujuhong replied:
Yes, it could be. I don't think there is much value of calibrating this now, but we can certainly do that later.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -110,6 +111,18 @@ const (
+	// Relisting is used to discover missing container events.
+	// Use a shorter period because generic PLEG relies on relisting for
+	// discovering container events.
+	plegRelistPeriod = time.Second * 3
```
This is an implementation detail. We can have a flag just for the stabilization phase maybe, but I'd prefer not cutting a release with the flag. 
Moreover, once we centralize all the pod interactions, the re-listing period should not be of concern right?

yujuhong replied:
I am okay with not exposing this as a flag. It is a bit similar to the resyncInterval, which we expose now though. 

Relisting period will always be a concern unless we adopt docker/rkt event stream completely. But even then, the relisting period would affect container runtimes that don't have an event stream.

yujuhong replied:
Removed the TODO.

vishh replied:
Why is it of concern? Can you elaborate? Once you add the specialized cache, are there any other issues ?

yujuhong replied:
Relisting is still required to populate the cache. The frequency of relisting affects how quick kubelet discovers container events (e.g., containers died recently).

vishh replied:
Acknowledged!

On Thu, Nov 12, 2015 at 2:10 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> In pkg/kubelet/kubelet.go
> <https://github.com/kubernetes/kubernetes/pull/13571#discussion_r44722406>
> :
>
> > @@ -110,6 +111,18 @@ const (
> >  	housekeepingPeriod = time.Second * 2
> >
> >  	etcHostsPath = "/etc/hosts"
> > +
> > +	// Capacity of the channel for recieving pod lifecycle events. This number
> > +	// is a bit arbitrary and may be adjusted in the future.
> > +	plegChannelCapacity = 1000
> > +
> > +	// TODO(yujuhong): migrate some of the parameters to kubelet flags.
> > +	// Relisting is used to discover missing container events.
> > +	// Use a shorter period because generic PLEG relies on relisting for
> > +	// discovering container events.
> > +	plegRelistPeriod = time.Second * 3
>
> Relisting is still required to populate the cache. The frequency of
> relisting affects how quick kubelet discovers container events (e.g.,
> containers died recently).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13571/files#r44722406>.
>

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -110,6 +111,18 @@ const (
+	// discovering container events.
+	plegRelistPeriod = time.Second * 3
+
+	backOffPeriod = time.Second * 10
```
nit: Document this value.

yujuhong replied:
Done.

yujuhong commented:
Running failed e2e tests locally. Will update results soon.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -2112,6 +2132,25 @@ func (kl *Kubelet) syncLoopIteration(updates <-chan kubetypes.PodUpdate, handler
 			// TODO: Do we want to support this?
 			glog.Errorf("Kubelet does not support snapshot update")
 		}
+	case e := <-plegCh:
```
We should drain the `plegCh` during every `syncLoopIteration` right? IIUC, this logic processes only one event as of now.

yujuhong replied:
This sounds like an optimization. I'd prefer doing that later if you don't mind. Right now, it seems pretty fair to get only one event from the channel, when there are multiple channels in `select`. We can consider draining and batching some of them in the future.

vishh replied:
I clarified the existing behavior offline with @yujuhong. No changes required. I take back my comment.

yujuhong commented:
@k8s-bot test this please

All failed tests passed locally.

vishh commented:
LGTM! Looking forward to the performance improvements from this PR.

wojtek-t commented:
@yujuhong can you please rebase?

yujuhong commented:
Rebased and re-applied lgtm

yujuhong commented:
@k8s-bot test this please

pmorie commented on cmd/kubelet/app/server.go:
```diff
@@ -216,7 +216,7 @@ func NewKubeletServer() *KubeletServer {
 		SerializeImagePulls:            true,
 		StreamingConnectionIdleTimeout: 5 * time.Minute,
-		SyncFrequency:                  10 * time.Second,
+		SyncFrequency:                  1 * time.Minute,
```
Seems like a pretty big change that is not strictly related to the point of this PR... am I missing something?

yujuhong replied:
The primary focus of this PR is to reduce redundant syncs (10 second periodic syncs) when there is no change. The added PLEG already checks the container runtime frequently for any container changes. We keep the periodic sync as a fallback, in case anything is missed, but it should be less frequent.

pmorie replied:
@yujuhong cool, that makes sense to me -- this seems like the kind of thing that could affect some of the E2Es that are outside the merge path -- have you run any of those?

yujuhong commented:
@k8s-bot test this please

pmorie commented on pkg/kubelet/kubelet.go:
```diff
@@ -111,6 +112,19 @@ const (
 
 	etcHostsPath = "/etc/hosts"
+
+	// Capacity of the channel for recieving pod lifecycle events. This number
```
Nit: receiving

yujuhong commented:
@k8s-bot test this please

yujuhong commented:
@k8s-bot test this please

yujuhong commented:
@k8s-bot test this please

yujuhong commented:
@k8s-bot test this please

yujuhong commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot e2e test this please

yujuhong commented:
@k8s-bot test this

yujuhong commented:
@k8s-bot test this

yujuhong commented:
FYI, I checked the failed test in [build](http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//16609), where `kubectl describe pod` was not able to get the relevant events for the pod in question. However, the corresponding kubelet.log recorded all the events and pod status generation correctly. I couldn't find any suspicious issues on the kubelet side.

huang195 commented:
@yujuhong is there a PR on having kubelet using the docker event stream instead of polling?

yujuhong commented:
> @yujuhong is there a PR on having kubelet using the docker event stream instead of polling?

Nope. It wasn't prioritized for v1.3, so no one is actively working on it. Also, it'd need most likely more than one PR:
 - `SyncPod` does too many things, so we'll need to ensure the expected events are properly recorded
 - `rkt` doesn't support event stream yet, so this will only work for docker.
